src/main/java/com/rocketden/main/service/ProblemService.java

```diff
@@ -83,6 +86,68 @@ public ProblemDto getProblem(String problemId) {
         return ProblemMapper.toDto(problem);
     }
 
+    public ProblemDto editProblem(String problemId, ProblemDto updatedProblem) {
+        Problem problem = repository.findProblemByProblemId(problemId);
+
+        if (problem == null) {
+            throw new ApiException(ProblemError.NOT_FOUND);
+        }
+
+        if (updatedProblem == null || updatedProblem.getName() == null
+                || updatedProblem.getDescription() == null
+                || updatedProblem.getDifficulty() == null
+                || updatedProblem.getProblemInputs() == null
+                || updatedProblem.getOutputType() == null) {
+            throw new ApiException(ProblemError.EMPTY_FIELD);
+        }
+
+        if (updatedProblem.getDifficulty() == ProblemDifficulty.RANDOM) {
+            throw new ApiException(ProblemError.BAD_DIFFICULTY);
+        }
+
+        // Ensure that the user entered valid inputs and outputs for the problem
+        for (ProblemTestCaseDto input : updatedProblem.getTestCases()) {
+            validateInputsGsonParseable(input.getInput(), updatedProblem.getProblemInputs());
+            validateGsonParseable(input.getOutput(), updatedProblem.getOutputType());
+        }
```

Comment:
This for loop checks for the test cases provided in the `updatedProblem`, but it does not check if this fits with the test cases already existing in the `problem` object in the database. This presents an issue where a user can create a problem that accepts an **ARRAY_INTEGER** input with the appropriate test case, and then the same user can send an edit request to the problem with an **INTEGER** input an a _new_ appropriate test case. Then, both test cases will exist in the database, where one of them does not fit the current problem schema. I see a few potential solutions to this:

1. Do not allow updates to the `problemInputs` and `outputType` fields. This would restrict editing capabilities but effectively remove this problem.
2. Allow updates to `problemInputs` and `outputType` fields, and remove the existing test cases that _conflict_ with this change (many input types can be seen as another, like STRING and INTEGER in certain cases).
3. Allow updates to `problemInputs` and `outputType` fields, and throw an error if _any_ test case (in existing or updated) conflict with the change.
4. Allow updates to `problemInputs` and `outputType` fields, then remove all existing test cases if these fields are updated.
5. Add a `boolean` value called `rewriteTests` to the `editProblem` request (preferably in the request body, though this would require a new object) that always removes any tests in the problem and replaces them with the new tests in `updatedProblem`.
6. Add a `boolean` value called `rewriteTests` to the `editProblem` request (preferably in the request body, though this would require a new object) that always removes any tests in the problem _if they conflict_ with any new schema that is set.

There are probably more solutions as well -- let me know if you have any preference!

Reply:
Good catch - I actually created a test case for this (`editProblemNewProblemInputInvalidatesTestCases`) along the lines of solution 3, but it's a good thing you brought this up because I looked at it again and there's a bug haha - the test cases in updatedProblem are supposed to contain all the test cases (old and new) so it should throw an error if any of the inputs get invalidated, but I forgot to clear/replace the existing test cases in the database before setting the new ones so it's basically just adding a whole bunch of duplicate test cases on every edit

It should be fixed now - lmk if you notice any other bugs and things like that

Reply:
Awesome, it's fixed on my end as well! To clarify, if the `updatedProblem` has no test cases while the `problem` has some old test cases, this means that the user is attempting the edit the problem and remove all test cases?

This sounds like a fine solution, I just wanted to make sure that's actually what is happening here.

Reply:
Yep - that's the intention